packages/plugin-edgedb/src/types.ts

```diff
@@ -0,0 +1,389 @@
+import type { Client } from 'edgedb';
+import {
+  FieldMap,
+  FieldNullability,
+  InputFieldMap,
+  InputShapeFromFields,
+  InterfaceParam,
+  MaybePromise,
+  Normalize,
+  ObjectRef,
+  SchemaTypes,
+  ShapeWithNullability,
+  TypeParam,
+} from '@pothos/core';
+import type { EdgeDBObjectFieldBuilder } from './edgedb-field-builder';
+import { GraphQLResolveInfo } from 'graphql';
+
+export interface EdgeDBQueryBuilder {
+  default: unknown;
+}
+export interface EdgeDBDriver extends Client {}
+
+export type EdgeDBModels<
+  Name extends EdgeDBTypeKeys,
+  Types extends SchemaTypes = SchemaTypes,
+> = Types['EdgeDBTypes']['default'][Name] extends infer T ? T : never;
+
+export type EdgeDBTypeKeys<Types extends SchemaTypes = SchemaTypes> =
+  keyof Types['EdgeDBTypes']['default'] extends infer Key
+    ? Key extends string
+      ? Key
+      : never
+    : never;
+
+export type EdgeDBDefaultExportKeyTypes<DefaultExports> = {
+  [K in keyof DefaultExports]-?: K extends string
+    ? string
+    : K extends number
+    ? number
+    : K extends symbol
+    ? symbol
+    : never;
+}[keyof DefaultExports];
+
+type EdgeDBDefaultExportKey<
+  DefaultExports,
+  KeyType extends string | number | symbol = EdgeDBDefaultExportKeyTypes<DefaultExports>,
+> = Extract<keyof DefaultExports, KeyType>;
+
+export type EdgeDBModelTypes<Types extends SchemaTypes> =
+  Types['EdgeDBTypes']['default'] extends infer ObjectTypeMap
+    ? ObjectTypeMap extends object
+      ? ObjectTypeMap
+      : never
+    : never;
+
+export type EdgeDBModelShape<
+  Types extends SchemaTypes,
+  Name extends EdgeDBTypeKeys<Types>,
+> = EdgeDBModelTypes<Types>[Name] extends infer Property
+  ? Property extends BaseObject & EdgeDBModelTypes<Types>[Name]
+    ? Property & {
+        Fields: extractLinks<Property> extends infer Field
+          ? Field extends string
+            ? SplitLT<Field>
+            : null
+          : never;
+        Links: {
+          [Key in Property['Fields']]: {
+            Shape: Property['Fields'][Key] extends infer Shape
+              ? Shape extends BaseObject
+                ? Shape
+                : Shape extends Array<BaseObject>
+                ? Shape[0]
+                : never
+              : never;
+          };
+        };
+      }
+    : never
+  : never;
+
+export type SelectedKeys<T> = { [K in keyof T]: T[K] extends false ? never : K }[keyof T];
+
+// --
+// EdgeDB Types
+// --
+export enum Cardinality {
+  AtMostOne = 'AtMostOne',
+  One = 'One',
+  Many = 'Many',
+  AtLeastOne = 'AtLeastOne',
+  Empty = 'Empty',
+}
+
+export enum TypeKind {
+  scalar = 'scalar',
+  // castonlyscalar = "castonlyscalar",
+  enum = 'enum',
+  object = 'object',
+  namedtuple = 'namedtuple',
+  tuple = 'tuple',
+  array = 'array',
+  range = 'range',
+}
+export type tupleOf<T> = [T, ...T[]] | [];
+export type cardinalityAssignable<C extends Cardinality> = C extends Cardinality.Empty
```

Comment:
this could probably be modeled in a way thats more readable:

```ts
type CardinalityAssignable<C extends Cardinality> =  {
  Empty: Cardinality.Empty
  ... 
  AtLeastOne: Cardinality.One | Cardinality.AtLeastOne | Cardinality.Many
}[C]